presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/InMemoryCachingHiveMetastore.java

```diff
@@ -373,7 +373,16 @@ public Optional<Table> getTable(MetastoreContext metastoreContext, String databa
     @Override
     public Optional<Table> getTable(MetastoreContext metastoreContext, HiveTableHandle hiveTableHandle)
     {
-        return get(tableCache, getCachingKey(metastoreContext, hiveTableHandle));
+        KeyAndContext<HiveTableHandle> key = getCachingKey(metastoreContext, hiveTableHandle);
+        Optional<Table> table = get(tableCache, key);
+        // If the table is non-existent, we store [Key=Table, Value=Optional.Empty] cache entry. If someone outside
+        // Presto creates the table, Presto keeps throwing "Table does not exist" until refresh time is reached. So,
+        // forcing a refresh for non-existent table once here.
+        if (!table.isPresent()) {
+            tableCache.refresh(key);
```

Comment:
nit: Just thinking out loud, should we invalidate the entry instead of refreshing it? While this may not lead to a significant improvement, invalidating the entry—rather than caching an empty table—could save an extra call to the metastore when the table is empty the very first time. Since we plan to call the metastore for any cached empty table, it might be more efficient not to cache it at all. This approach would also provide more accurate JMX metrics. If we continue to cache non-existent tables, the count of actual cached tables may not be accurate since these empty table entries would also be counted.

Reply:
I am aligned with this.